DDF-1960 merged catalog and content frameworks

stustison opened the pull request:
#### What does this PR do?
Removes the DDF Content application and merges the functionality into DDF Catalog
#### Who is reviewing it?
@coyotesqrl @lessarderic @pklinef @kcwire @jlcsmith @Lambeaux @brendan-hofmann 
#### How should this be tested?
run through all itests
start up DDF, ingest a file using the Search UI ingest, perform a query and see that the result exists, download the product and see that the files match
ingest a metacard XML document by posting the document as the body of the message, see that a metacard result exists for this document and that no product is available
#### Any background context you want to provide?
Additional I-tests will be added in a separate PR as there were none existing for most of DDF Content
#### What are the relevant tickets?
DDF-1960
#### Screenshots (if appropriate)
#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

 - removed content app
 - rewrote file system storage provider to temporarily store products until commit
 - updated CatalogFramework API with new create and update methods
 - updated Search UI to point to catalog REST endpoint
 - updated catalog REST endpoint to use new create and update methods for multipart requests

lessarderic commented:
**`CatalogFrameworkImpl.create(CreateStorageRequest streamCreateRequest)`**

1. L878: Can’t we just check if `streamCreateRequest.getContentItems().size() == 0` at the top of the method and return right away if that’s the case?
2. L908: Any chance that one of the `contentItem` be `null`? How about `contentItem.getId()`?
3. L918: I believe this `if (createStorageRequest != null)` won't be needed anymore if comment #1 is addressed.
4. L917: Should we have some kind of `catch (RuntimeException)` here and rollback any changes we've made, or is it ok to commit only some of the content items if something goes wrong?

Lambeaux commented on catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/content/data/impl/ContentItemImpl.java:
```diff
@@ -0,0 +1,162 @@
+                this.mimeType = new MimeType(this.mimeTypeRawData);
+            } catch (MimeTypeParseException e) {
+                LOGGER.debug("Unable to create MimeType from raw data " + this.mimeTypeRawData);
+            }
```
Redundant `try` blocks - I would move the entire `if` statement into a single `try-catch`.

stustison replied:
fixed

lessarderic commented:
**`CatalogFrameworkImpl.generateMetacard()`**

1. L661: Shouldn't we create `causeMessage` in the catch block instead?
2. L670: Rename iterable to `candidate` instead of `listOfCandidate`?
3. L682: Can we surround this with a `if (LOGGER.isDebugEnabled())` and use `String.format("Could not create metacard with mimeType %s. Reason: %n%s, mimeTypeRaw, e.getMessage())` to initialize the `StringBuilder` instead of making all those `append` calls? We could at the very least chain the `append` calls.
4. L703: Can metacard IDs ever be `null`?
5. L723: Don't think we need to check whether `subject` is `null` here. If it's `null`, `SubjectUtils.getName(subject)` will return `null`, which should be ok when we get to line 728.

stustison commented:
@lessarderic 
first block
1. fixed
2. it won't be null from any of our impls
3. fixed
4. fixed
second block
1. if you do that, you can't use it in the next block down
2. fixed
3. fixed
4. metacard ids can be null when you're creating a metacard for the first time, so for create it will always be null and for update it will always be non-null
5. fixed

lessarderic commented on catalog/core/catalog-core-api/src/main/java/ddf/catalog/CatalogFramework.java:
```diff
@@ -70,14 +72,43 @@
+     * <p>
+     * <b>Implementations of this method must:</b>
+     * <ol>
+     * <li/>Before evaluation, call {@link ddf.catalog.plugin.PreIngestPlugin#process(CreateRequest)} for each
```
Wrong request class name and pre and post plugin names used in the comments.

stustison replied:
These are actually correct, the storage request does not go through this path.

lessarderic commented on catalog/core/catalog-core-api/src/main/java/ddf/catalog/CatalogFramework.java:
```diff
@@ -70,14 +72,43 @@
+     * <li/>Before evaluation, call {@link ddf.catalog.plugin.PreIngestPlugin#process(CreateRequest)} for each
+     * registered {@link ddf.catalog.plugin.PreIngestPlugin} in order determined by the OSGi SERVICE_RANKING
+     * (Descending, highest first), "daisy chaining" their responses to each other.
+     * <li/>Call {@link ddf.catalog.source.CatalogProvider#create(CreateRequest)} on the registered
```
Don't think we're calling the `CatalogProvider` here. Wrong response class name as well.

stustison replied:
These are correct as well.

lessarderic replied:
I'm confused. This method is taking in a `CreateStorageRequest`, so I would expect the implementations to call the `PreCreateStoragePlugin` plugins, not the `PreIngestPlugin` ones, then call the `StorageProvider`, not `CatalogProvider`, and finally call the `PostCreateStoragePlugin`.

What am I missing?

stustison replied:
Those plugins are called, but they are called after all of the original catalog plugins. That also includes the security plugins so that the metadata itself can be evaluated before any of the storage operations are executed.

lessarderic commented on catalog/core/catalog-core-api/src/main/java/ddf/catalog/CatalogFramework.java:
```diff
@@ -416,10 +447,46 @@ public BinaryContent transform(Metacard metacard, String transformerId,
+     * <p>
+     * <b>Implementations of this method must:</b>
+     * <ol>
+     * <li/>Before evaluation, call {@link ddf.catalog.plugin.PreIngestPlugin#process(UpdateRequest)} for each
```
Same comment as above. Need to make sure these comments are updated.

lessarderic commented:
**`CatalogFrameworkImpl.guessFileExtension()`**
1. Can we rename  to something like `updateFileExtension()` or `getFileNameWithMimeTypeExtension()`? Would make it clearer that the method actually returns the full file name, not just the file extension.
2. L740: It's usually bad practice to reassign to a parameter

**`CatalogFrameworkImpl.guessMimeType()`**
1. L759: Reassignment to a parameter

**`CatalogFrameworkImpl.update(UpdateStorageRequest)`**
1. There is a lot of duplication between this method and `create(CreateStorageRequest)` and most of the comments I made on the former apply here as well.
2. L1154: We should be able to make that check at the top and just return right away. Actually, looks like `validateUpdateStorageRequest()` is already checking for this, so we should be able to get rid of this check (same in `create(CreateStorageRequest)` method.
3. L1187: Do we care about duplicate IDs?
4. L1203: Rename exception variable to `e` instead of `e1` and pass as argument to `LOGGER.error()`. Actually, should this be an error or a warning?
5. L1219: Same comment

**`CatalogFrameworkImpl.delete(DeleteRequest)`**
1. L1473: Pass `e` to `InternalIngestException` ctor. Also, shouldn't we log an error instead of a warning?

**`CatalogFrameworkImpl`**
1. L2052, 2081: Can we build these strings using `String.format()` instead?

lessarderic commented:
**`Constants`**
1. New `MIME_TYPE` constant doesn't seem to be used anywhere. Can we remove? `Constant` classes are an anti-pattern anyways so we really shouldn't be adding anything else to this class.

**`ddf.catalog.content.operation.*`**
1. Can we add something in those interface comments that say whether a returned value can be `null` or not?
2. Can we also explain what the IDs are used for?

**`ContentItemImpl`**
1. Any reason all fields are `protected` instead of being `private`? Especially strange considering that they all have getters.
2. Comments need to be reformatted and reworded (weird sentences) and missing ones need to be added (e.g., third ctor)
3. Since the class assumes that `filename` must be set before a `File` object can be returned, I don't think L118 makes sense. Also, to avoid any discrepancies between the `filename` and the `File` object, the `file` field _really_ needs to be `private`. If we add a setter for `file`, the class needs to make sure both `file` and `filename` stay in-sync.

stustison commented:
phil
another API suggestion, all the storage requests should extend from a single interface that includes the ID, then the commit and rollback methods should take the original request instead of a String ID

Lambeaux commented:
Code all clear! QA testing passed with no issues uploading and querying for products of various sizes and formats. Up-to-date full build passed! LGTM :+1:

lessarderic commented:
**`CreateStorageRequestImpl`**
1. L30: Don't need to create the `ArrayList` here
2. Need JavaDoc comments on that class and ctors
3. Should we have a setter for the content items or keep the object immutable?

**`DeleteStorageRequestImpl`**
1. L30: Don't need to create the `ArrayList` here
2. Need JavaDoc comments on that class and ctors

**`ReadStorageRequestImpl`**
1. Need JavaDoc comments on that class and ctors

**`UpdateStorageRequestImpl`**
1. Most of the code here is the same as `CreateStorageRequestImpl`. We should consider creating a base class.

**`ddf.catalog.content.operation.impl`**
1. All the `ResponseImpl` classes share the same exact code. They should all extend from a common base class.

lessarderic commented:
**`ContentResourceReader`**
1. We should be able to replace the static initializer with `private static Set<String> qualifierSet = Collections.singleton(URL_CONTENT_SCHEME)`
2. Should the ctor need check for a `null` list?
3. May want to add a comment on the `bind` and `unbind` method saying that those are called by blueprint and what the expected parameter values are.

**`StorageProviderImpl`**
1. JavaDoc comments on class please :-) Or give the class a better name, e.g., `FileBasedStorageProvider` or something.
2. L109: Assumption here is that the caller will log something?
3. L130: Reverse the condition and do an early throw instead instead of having everything inside the `if` block.
3. L138: Can we catch something more specific than `Exception` here? Also, if we fail because we get an `IOException` when `Files.newInputStream(file)` is called, shouldn't we just exit the method instead of keep going?
4. L161: Any reason we're using a FQN instead of importing `com.google.common.io.Files`?
5. If the admin changes the content directory, none of the existing content item files will be moved, right? That means that you could potentially have more than one directory that contains content items. Do we _really_ want to allow admins to change that directory _after_ they ingested content items? If we do, it makes it almost impossible to do provider-level migration of the data since you can't simply copy a single directory anymore, i.e., you'd have to remember _all_ the directories where content items were ever stored.
6. L241: Reverse condition and early return instead of putting everything inside the `if` block. Should we log something too?
7. We should extract the code that deals with the directory structure to a different class. That way we could make it a little more generic and potentially reusable if needed. We could also make the implementation vary in the future if need be. For instance, we could have the current implementation that uses the first 6 characters of the metacard ID to distribute the files between directories, but be able to substitute the implementation to something that distributes them some other way based on how the IDs are formatted.
8. To be on the safe side, can we add a check in `getContentFilePathParts()` to make sure the metacard ID has at least 6 characters? The provider shouldn't fail just for that reason.

brianfelix commented:
:+1:
